referenceframe/ptgFrame.go

```diff
@@ -0,0 +1,166 @@
+package referenceframe
+
+import (
+	"fmt"
+	"math"
+
+	pb "go.viam.com/api/component/arm/v1"
+
+	"go.viam.com/rdk/motionplan/tpspace"
+	"go.viam.com/rdk/spatialmath"
+)
+
+const (
+	defaultSimDistMM      = 1100.
+	defaultAlphaCnt  uint = 121
+)
+
+const (
+	ptgIndex int = iota
+	trajectoryIndexWithinPTG
+	distanceAlongTrajectoryIndex
+)
+
+type ptgFactory func(float64, float64, float64) tpspace.PrecomputePTG
+
+var defaultPTGs = []ptgFactory{
+	tpspace.NewCirclePTG,
+	tpspace.NewCCPTG,
+	tpspace.NewCCSPTG,
+	tpspace.NewCSPTG,
+	tpspace.NewAlphaPTG,
+}
+
+type ptgGridSimFrame struct {
+	name       string
+	limits     []Limit
+	geometries []spatialmath.Geometry
+	ptgs       []tpspace.PTG
+}
+
+// NewPTGFrameFromTurningRadius will create a new Frame which is also a tpspace.PTGProvider. It will precompute the default set of
+// trajectories out to a given distance, or a default distance if the given distance is <= 0.
+func NewPTGFrameFromTurningRadius(name string, velocityMMps, turnRadMeters, simDist float64, geoms []spatialmath.Geometry) (Frame, error) {
+	if velocityMMps <= 0 {
+		return nil, fmt.Errorf("cannot create ptg frame, movement velocity %f must be >0", velocityMMps)
+	}
+	if turnRadMeters <= 0 {
+		return nil, fmt.Errorf("cannot create ptg frame, turning radius %f must be >0", turnRadMeters)
+	}
+
+	if simDist <= 0 {
+		simDist = defaultSimDistMM
+	}
+
+	// Get max angular velocity in radians per second
+	maxRadps := velocityMMps / (1000. * turnRadMeters)
+	pf := &ptgGridSimFrame{name: name}
+	err := pf.initPTGs(velocityMMps, maxRadps, simDist)
+	if err != nil {
+		return nil, err
+	}
+
+	pf.geometries = geoms
+
+	pf.limits = []Limit{
+		{Min: 0, Max: float64(len(pf.ptgs) - 1)},
+		{Min: 0, Max: float64(defaultAlphaCnt)},
+		{Min: 0},
+	}
+
+	return pf, nil
+}
+
+func (pf *ptgGridSimFrame) DoF() []Limit {
+	return pf.limits
+}
+
+func (pf *ptgGridSimFrame) Name() string {
+	return pf.name
+}
+
+// TODO: Define some sort of config struct for a PTG frame.
+func (pf *ptgGridSimFrame) MarshalJSON() ([]byte, error) {
+	return nil, nil
+}
+
+// Inputs are: [0] index of PTG to use, [1] index of the trajectory within that PTG, and [2] distance to travel along that trajectory.
+func (pf *ptgGridSimFrame) Transform(inputs []Input) (spatialmath.Pose, error) {
+	ptgIdx := int(math.Round(inputs[ptgIndex].Value))
+	trajIdx := uint(math.Round(inputs[trajectoryIndexWithinPTG].Value))
+	traj := pf.ptgs[ptgIdx].Trajectory(trajIdx)
+	lastPose := spatialmath.NewZeroPose()
+	for _, trajNode := range traj {
+		// Walk the trajectory until we pass the specified distance
+		if trajNode.Dist > inputs[distanceAlongTrajectoryIndex].Value {
```

Comment:
I think this is cleaner

```
var trajNode [VARIABLE_TYPE]
for _, trajNode = range traj {
  if trajNode.Dist <= inputs[distanceAlongTrajectoryIndex].Value {
    break
  }
}
return trajNode.Pose
```

Reply:
Also I think we've "walked the trajectory" a couple times in this PR.  Is it worth putting this in its own function?